# PIPE
Protein-protein interaction prediction engine

Code for PIPE as described in the following publications:

* Pitre S, Dehne F, Chan A, Cheetham J, Duong A, Emili A, Gebbia M, Greenblatt J, Jessulat M, Krogan N, Luo X, Golshani A (2006) BMC Bioinformatics 7:365
* Pitre S, Alamgir Md, Green JR, Dumontier M, Dehne F, Golshani A, 2008, "Computational Methods for Predicting Protein-Protein Interactions", Adv Biochem Eng Biotechnol. 110:247-267. (Review)
* Pitre S, North C, Alamgir Md, Jessulat M, Chan A, Luo X, Green JR, Dumontier M, Dehne F, Golshani A, 2008, "Global investigation of protein-protein interactions in yeast Saccharomyces cerevisiae using re-occurring short polypeptide sequences", Nucleic Acids Research 36(13):4286-4294.
* Jessulat M, Pitre S, Gui Y, Hooshyar M, Omidi O, Samanfar B, Tan LH, Alamgir Md, Green JR, Dehne F, Golshani A, 2011, "Recent Advances in Protein-Protein Interaction Prediction: Experimental and Computational Methods", Expert Opinion on Drug Discovery, 6(9):921-935 (doi:10.1517/17460441.2011.603722) (Review)
* Amos-Binks A, Patulea C, Pitre S, Schoenrock A, Gui Y, Green JR, Golshani A, Dehne F, 2011, "Binding Site Prediction for Protein-Protein Interactions and Novel Motif Discovery using Re-occurring Polypeptide Sequences", BMC Bioinformatics, 12:225. 
* Pitre S, Hooshyar M, Schoenrock A, Samanfar B, Jessulat M, Green JR, Dehne F, Golshani A, 2012, "Short Co-occurring Polypeptide Regions Can Predict Global Protein Interaction Maps", Scientific Reports, 2:239.
* Schoenrock A, Samanfar B, Pitre S, Hooshyar M, Jin K, Phillips CA, Wang H, Phanse S, Omidi K, Gui Y, Alamgir Md, Wong A, Barrenäs F, Babu M, Benson M, Langston MA, Green JR, Dehne F, Golshani A, 2014, "Efficient Prediction of Human Protein-Protein Interactions at a Global Scale", BMC Bioinformatics, 15:283.
* Schoenrock A, Burnside D, Moteshareie H, Pitre S, Hooshyar M, Green JR, Golshani A, Dehne F, Wong A, 2017, "Evolution of Protein-Protein Interaction Networks in Yeast," PLoS One 12(3): e0171920 (doi:10.1371/journal.pone.0171920).
* Kazmirchuk T, Dick K, Burnside DJ, Barnes B, Moteshareie H, Hajikarimlou M, Omidi K, Ahmed D, Low A, Lettl C, Hooshyar M, Schoenrock A, Pitre S, Babu M, Cassol E, Samanfar B, Wong A, Dehne F, Green JR, Golshani A, 2017, "Designing Anti-Zika Virus Peptides Derived from Predicted Human-Zika Virus Protein-Protein Interactions," Computational Biology and Chemistry, 71:180-197 (doi:10.1016/j.compbiolchem.2017.10.011).
* Grigg N, Schoenrock A, Dick K, Green JR, Golshani A, Wong A, Dehne F, Tsai EC, Biggar KK, 2019, "Insights into the suitability of utilizing brown rats (Rattus norvegicus) as a model for healing spinal cord injury with epidermal growth factor and fibroblast growth factor-II by predicting protein-protein interactions," Computers in Biology and Medicine, 104:220-226, doi:10.1016/j.compbiomed.2018.11.026.
*  Burnside D, Schoenrock A, Moteshareie H, Hooshyar M, Basra P, Hajikarimloo M, Dick K, Barnes B, Kazmirchuk T, Jessulat M, Pitre S, Samanfar B, Babu M, Green JR, Wong A, Dehne F, Biggar KK, Golshani A, 2019, "A robust computational tool for engineering synthetic binding proteins from random amino acid sequences," ISCIENCE 11:375-387, doi:10.1016/j.isci.2018.11.038.
* Dick K, Samanfar B, Barnes B, Cober ER, Mimee B, Tan LH, Molnar SJ, Biggar KK, Golshani A, Dehne F, Green JR, 2020, "PIPE4: Fast PPI Predictor for Comprehensive Inter- and Cross-Species Interactomes", Scientific Reports, 10:1390, doi:10.5683/SP2/PVOTRN.
* Dick K, Hooker J, Nissan N, Pattang A, Sadowski M, Barnes B, Tan LH, Burnside D, Phanse S, Aoki H, Babu M, Dehne F, Golshani A, Cober ER, Green JR, Samanfar B, 2021, "Human-Soybean Allergies: Elucidation of the Seed Proteome & Comprehensive Protein-Protein Interaction Prediction," Journal of Proteome Research, 20(11):4925–4947, doi:10.1021/acs.jproteome.1c00138.

The "master" branch created by Brad Barnes with code from his thesis:
Bradley D Barnes, "Accelerated Transfer Learning for ProteinProtein Interaction Prediction", 2018, Carleton University
https://curve.carleton.ca/e61aef98-ca4e-4643-ae4b-fc33a498bd05

Most recent code is likely in the "Deep-Pipe-Sites" branch created by William Ma and Thomas Gatto as part of their SYSC4907 4th-year project in 2020-21. That branch may include code for interaction _site_ determination, beyond PIPE.

This code is unsupported and may be quite difficult to use.
